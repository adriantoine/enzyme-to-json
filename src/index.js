import omit from 'lodash.omit';
import pickBy from 'lodash.pickby';
import compact from 'lodash.compact';
import get from 'lodash.get';

export function shallowToJson(wrapper) {
    const type = wrapper.type();
    if (!type) {
        return wrapper.node;
    }

    const children = compact(wrapper.children().map(c => shallowToJson(c)));
    const props = pickBy(wrapper.props(), val => val !== undefined);

    const json = {
        type: wrapper.name(),
        props: omit(props, 'children'),
        children: (children.length) ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };

    // If the type of element is not a function, it means that is a DOM element
    if (typeof type !== 'function') {
        return json;
    }

    // We need to get the rendered component in Enzyme internals
    // because it won't be returned by `.children()`
    const element = get(wrapper, 'node._reactInternalInstance._renderedComponent._currentElement');
    if(!element) {
        return json;
    }

    const jsonChildren = json.children;
    json.children = [{
        type: element.type,
        props: omit(element.props, 'children'),
        children: jsonChildren,
        $$typeof: Symbol.for('react.test.json'),
    }];

    return json;
}

export {shallowToJson as mountToJson};

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};
export const renderToJson = wrapper =>
    renderChildToJson(wrapper.children()[0]);
