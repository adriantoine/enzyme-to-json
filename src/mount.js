import compact from 'lodash.compact';
import omitBy from 'lodash.omitby';
import values from 'object-values';
import {isDOMComponent, isElement} from 'enzyme/build/react-compat';
import {internalInstance, propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';

function instToJson(inst) {
    if (typeof inst === 'string' || typeof inst === 'number') return inst;
    if (!inst) return '';

    if (!inst.getPublicInstance) {
        const internal = internalInstance(inst);
        return instToJson(internal);
    }
    const publicInst = inst.getPublicInstance();

    if (typeof publicInst === 'string' || typeof publicInst === 'number') return publicInst;
    if (!publicInst && !inst._renderedComponent) return '';

    const currentElement = inst._currentElement;
    const type = typeName(currentElement);
    const props = omitBy(propsOfNode(currentElement), (val, key) => key === 'children' || val === undefined);
    const children = [];
    if (isDOMComponent(publicInst)) {
        const renderedChildren = inst._renderedChildren;
        if (!renderedChildren) {
            children.push(...childrenOfNode(currentElement));
        } else {
            children.push(...values(renderedChildren));
        }
    } else if (
        isElement(currentElement) &&
        typeof currentElement.type === 'function'
    ) {
        children.push(inst._renderedComponent);
    }

    const childrenArray = compact(children.map(n => instToJson(n)));

    return {
        type,
        props,
        children: childrenArray.length ? childrenArray : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function mountToJson(wrapper) {
    return instToJson(wrapper.node);
}
