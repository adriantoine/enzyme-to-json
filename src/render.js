import compact from 'lodash.compact';

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};

export default wrapper => renderChildToJson(wrapper.children()[0]);
