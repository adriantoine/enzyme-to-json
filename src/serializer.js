import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

import toJson from './';

module.exports = {
    test(wrapper) {
        return wrapper instanceof ShallowWrapper || wrapper instanceof ReactWrapper || wrapper.cheerio;
    },
    print(wrapper) {
        return toJson(wrapper);
    },
};
