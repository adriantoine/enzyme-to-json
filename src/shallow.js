import compact from 'lodash.compact';
import isPlainObject from 'lodash.isplainobject';
import omit from 'lodash.omit';
import entries from 'object.entries';
import {propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';
import {isElement} from 'enzyme/build/react-compat';

function nodeToJson(node) {
    if(!node) return node;

    if (typeof node === 'string' || typeof node === 'number') {
        return node;
    }

    if (!isElement(node)) {
        if (Array.isArray(node)) {
            return node.map(nodeToJson);
        }

        if (isPlainObject(node)) {
            return entries(node).reduce((obj, [key, val]) => {
                obj[key] = nodeToJson(val);
                return obj;
            }, {});
        }

        return node;
    }

    const children = compact(childrenOfNode(node).map(n => nodeToJson(n)));
    const type = typeName(node);
    const props = omit(propsOfNode(node), 'children');

    return {
        type,
        props: nodeToJson(props),
        children: children.length ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function shallowToJson(wrapper) {
    return nodeToJson(wrapper.node);
}
