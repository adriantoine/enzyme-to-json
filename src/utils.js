import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

const SHALLOW_WRAPPER_NAME = ShallowWrapper.name;
const REACT_WRAPPER_NAME = ReactWrapper.name;

export const isShallowWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === SHALLOW_WRAPPER_NAME;
export const isReactWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === REACT_WRAPPER_NAME;
export const isCheerioWrapper = wrapper => wrapper.cheerio;

export const isEnzymeWrapper = wrapper => isShallowWrapper(wrapper) || isReactWrapper(wrapper) || isCheerioWrapper(wrapper);
