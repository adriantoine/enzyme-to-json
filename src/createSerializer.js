import {isEnzymeWrapper} from './utils';
import toJson from './';

export default function createSerializer(options) {
  return {
    test(wrapper) {
      return isEnzymeWrapper(wrapper);
    },
    print(wrapper, serializer) {
      return serializer(toJson(wrapper, options));
    },
  };
}
