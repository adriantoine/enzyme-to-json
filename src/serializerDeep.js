import {isEnzymeWrapper} from './utils';
import toJson from './';

export function test(wrapper) {
  return isEnzymeWrapper(wrapper);
}

export function print(wrapper, serializer) {
  return serializer(toJson(wrapper, {mode: 'deep'}));
}
