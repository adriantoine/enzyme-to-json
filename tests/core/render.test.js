/* eslint-env jest */

import React from 'react';
import {render} from 'enzyme';
import {renderToJson} from '../../src';
import {BasicPure, BasicWithAList} from './fixtures/pure-function';
import {BasicClass, ClassWithNull} from './fixtures/class';

it('converts basic pure render', () => {
  const rendered = render(<BasicPure className="pure"><strong>Hello!</strong></BasicPure>);

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('converts basic class render', () => {
  const rendered = render(<BasicClass className="class"><strong>Hello!</strong></BasicClass>);

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('renders the whole list', () => {
  const wrapper = render(<BasicWithAList />);
  expect(renderToJson(wrapper.find('ul'))).toMatchSnapshot();
});

it('handles a component which returns null', () => {
  const rendered = render(<ClassWithNull />);

  expect(renderToJson(rendered)).toMatchSnapshot();
});

it('renders multiple elements as a result of find', () => {
  const rendered = render(<BasicWithAList />);
  expect(renderToJson(rendered.find('li'))).toMatchSnapshot();
});
