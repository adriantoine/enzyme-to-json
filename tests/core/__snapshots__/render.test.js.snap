exports[`test converts basic class render 1`] = `
<div
  class="basic-class class">
  <div
    class="group"
    id="group-id">
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      class="empty" />
  </div>
</div>
`;

exports[`test converts basic pure render 1`] = `
<div
  class="basic-pure pure">
  <div
    class="group"
    id="group-id">
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      class="empty" />
  </div>
</div>
`;

exports[`test handles a component which returns null 1`] = `null`;

exports[`test renders multiple elements as a result of find 1`] = `
Array [
  <li>
    0
</li>,
  <li>
    1
</li>,
]
`;

exports[`test renders the whole list 1`] = `
<ul>
  <li>
    0
  </li>
  <li>
    1
  </li>
</ul>
`;
